src/puppetlabs/puppetdb/query/paging.clj

```diff
@@ -152,43 +179,70 @@
         (dissoc :include_total)
         (assoc :count? count?))))
 
+(defn parse-int'
+  "Parses the int unless it's already an int"
+  [i]
+  (if (string? i)
+    (parse-int i)
+    i))
+
 (defn validate-limit
   "Validates that the limit string is a positive non-zero integer. Returns the integer
   form if validation was successful, otherwise an IllegalArgumentException is thrown."
   [limit]
-  {:pre  [(string? limit)]
+  {:pre  [(or (string? limit)
+              (integer? limit))]
    :post [(and (integer? %) (> % 0))]}
-  (let [l (parse-int limit)]
+  (let [l (parse-int' limit)]
     (if ((some-fn nil? neg? zero?) l)
       (throw (IllegalArgumentException.
               (format "Illegal value '%s' for :limit; expected a positive non-zero integer." limit)))
       l)))
 
+(pls/defn-validated parse-limit' :- (s/maybe s/Int)
+  "Parse the optional `limit` query parameter. Returns an integer
+  version of `limit` upon successful validation. Throws an exception
+  if provided `limit` is not a positive integer"
+  [limit :- (s/maybe (s/either String s/Int))]
+  (when limit (validate-limit limit)))
+
 (defn parse-limit
-  "Parse the optional `limit` query parameter in the paging options map,
+  "DEPRECATED - this function will be replaced by parse-limit'
+
+  Parse the optional `limit` query parameter in the paging options map,
   and return an updated map with the correct integer value.
   Throws an exception if the provided limit is not a positive non-zero integer."
+  {:deprecated "3.0.0"}
   [paging-options]
-  (update-in paging-options [:limit] #(when-not (nil? %) (validate-limit %))))
+  (update-in paging-options [:limit] parse-limit'))
```

Comment:
though somewhat more than 50% of me believes that leaving it the way it is is better, in this commit.